src/Payer.sol

```diff
@@ -21,83 +21,147 @@ import { Ownable } from 'openzeppelin-contracts/contracts/access/Ownable.sol';
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
     /**
      * @param account the account to send or mint to.
      * @param amount the amount of tokens `account` should receive, in {paymentToken} decimals.
      */
-    function sendOrMint(address account, uint256 amount) external onlyOwner {
+    function sendOrRegisterDebt(address account, uint256 amount) external onlyOwner {
         uint256 paymentTokenBalance = paymentToken.balanceOf(address(this));
 
         if (amount <= paymentTokenBalance) {
             paymentToken.safeTransfer(account, amount);
         } else if (paymentTokenBalance > 0) {
             paymentToken.safeTransfer(account, paymentTokenBalance);
-            iouToken.mint(account, amount - paymentTokenBalance);
+            registerDebt(account, amount - paymentTokenBalance);
         } else {
-            iouToken.mint(account, amount);
+            registerDebt(account, amount);
         }
     }
 
     function withdrawPaymentToken() external onlyOwner {
-        paymentToken.safeTransfer(owner(), paymentToken.balanceOf(address(this)));
+        address to = owner();
+        uint256 amount = paymentToken.balanceOf(address(this));
+        paymentToken.safeTransfer(to, amount);
+
+        emit TokensWithdrawn(to, amount);
     }
 
     /**
-     * @notice Redeem `account`'s IOU tokens in exchange for `paymentToken` in a best-effort approach, meaning it will
-     * attempt to redeem as much as possible up to `account`'s IOU balance, without reverting even if the amount is zero.
-     * Any account can redeem on behalf of `account`.
-     * @dev this function burns the IOU token balance that gets exchanged for `paymentToken`.
-     * @param account the account whose IOU tokens to redeem in exchange for `paymentToken`s.
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+      EXTERNAL FUNCTIONS (PUBLIC)
+     ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
      */
-    function redeem(address account) external {
-        uint256 amount = Math.min(iouToken.balanceOf(account), paymentToken.balanceOf(address(this)));
-        _redeem(account, amount);
-    }
 
-    function redeem(address account, uint256 amount) external {
-        _redeem(account, amount);
+    /// @notice Pays back debt up to `amount` of `paymentToken`
+    /// @param amount The maximum amount of tokens to send. This is expected to be the token balance of this contract
+    function payBackDebt(uint256 amount) external {
+        while (amount > 0 && !queue.empty()) {
+            DebtQueue.DebtEntry storage debt = queue.front();
+
+            uint256 _debtAmount = debt.amount;
+            address _debtAccount = debt.account;
+
+            if (amount < _debtAmount) {
+                // Not enough to cover entire debt, pay what you can and leave
+                debt.amount -= amount; // update remaining debt
+                totalDebt -= amount;
+                paymentToken.safeTransfer(_debtAccount, amount);
+                emit PaidBackDebt(_debtAccount, amount, false);
```

Comment:
could be nicer UX for event users to emit the remaining debt instead of a boolean, wdyt?

Reply:
good idea. done 32f6ce855ce99b4254e295132ed975f327daa0f4